### ISSUE TITLE
UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 78552: character maps to <undefined>

### ISSUE THREAD

**Issue by noklam** [quantumblacklabs/kedro#493]
## Description
pandas.CSVDataset default argument hit Unicode Error
```
UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 78552: character maps to <undefined>
```

## Steps
Step to reproduce

catalog.yml
```yaml
sample:
  type: pandas.CSVDataSet
  filepath: data2/sample.csv
```
python script to create and load dataset
```python
df['test'] = ['123']
df['test'] = '\x9d'

df.to_csv('sample.csv') # to data/sample.csv
pd.read_csv('sample.csv') # success
catalog.load('sample') # UnicodeDecodeError: 'charmap'
```

Kedro: 0.16.4
Window 10

**Comment by mzjp2** [comment]
Potentially related: https://github.com/quantumblacklabs/kedro/issues/291

**Comment by noklam** [comment]
Not sure if it is identical, I use a plain pd.read_csv() will read the files normally

**Comment by MerelTheisenQB** [comment]
Hi @noklam,

I think this might indeed be related to https://github.com/quantumblacklabs/kedro/issues/291. Could you try and see if the solution provided there solves your issue? 

"Try adding the following config to your catalog entry:
 
 ```
 fs_args:
     open_args_load:
         mode: "rb"
 ```
 
 We use `fsspec` underneath and it needs to open the file in binary mode. Try passing `encoding` to `open_args_load` or `load_args` as well, and that should work."

**Comment by noklam** [comment]
I pass Utf-8 to both load args and open_args_load and it work, I do not know why though.

I tried to debug with load args only and the encoding argument is passed to pandas successfully, not sure what is causing the issue.

**Comment by noklam** [comment]
Step to reproduce


catalog.yml
```yaml
sample:
  type: pandas.CSVDataSet
  filepath: data2/sample.csv
```
python script to create and load dataset
```python
df['test'] = ['123']
df['test'] = '\x9d'

df.to_csv('sample.csv') # to data/sample.csv
pd.read_csv('sample.csv') # success
catalog.load('sample') # UnicodeDecodeError: 'charmap'
```

**Comment by limdauto** [comment]
I think this is a discrepancy in pandas' API, not Kedro.

When writing data, the `pandas.CSVDataSet` uses panda's [`to_csv`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_csv.html) method. This method uses `utf-8` as the default encoding.

On the other hand, when reading data,  the `pandas.CSVDataSet` uses panda's [`read_csv`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) method. The subtle difference here is that this method has an _optional_ encoding, instead of an `utf-8` default. If it's not provided, it will defer to the encoding of the `open()` method on the filesystem, which will use [the system's default encoding](https://stackoverflow.com/a/36316351/1684058). On some Windows system, it's `cp1252`, hence the `charmap` error message.

You can fix this by either changing the encoding on the reading side or the writing side. Both are equally valid.

* On the reading side, you need to pass `encoding` to both `load_args`, which will then pass it to panda's `read_csv`, and `open_args_load`, which will then pass it to the `open()` method, so both use utf-8 correctly.
* Or on the writing side, you can pass `cp1252` to `save_args` so pandas writes your file with your system's default encoding and no need to adjust the load args.

I hope this helps.

**Comment by noklam** [comment]
What you have said suggested pd.read_csv do not use utf8. But my sample code actually just use plain read_csv and it works. Am I missing something?
thanks

**Comment by limdauto** [comment]
Ah, sorry, I think I got a bit confused as well. The opening of the file happens in the CSVDataSet, not pandas, so I guess you are right. In the dataset, we open the file [first](https://github.com/quantumblacklabs/kedro/blob/master/kedro/extras/datasets/pandas/csv_dataset.py#L155) before reading it. This opening doesn't use `utf-8` but rather the system's default. So can I suggest you try just putting `utf-8` in `open_args_load`, which when then be passed to the method, and see if it helps?

**Comment by noklam** [comment]
I test a few options
1. Success
```
fs_args:
    open_args_load:
        mode: "rb"
```
2. Success
```
fs_args:
    open_args_load:
        encoding: 'utf-8'
load_args:
  encoding: 'utf-8'
```
3.
```
load_args:
  encoding: 'utf-8'
```
failed
4. To test if kedro did use both encoding, 
```
fs_args:
    open_args_load:
        encoding: 'utf-8'
load_args:
  encoding: 'utf-1234'
```

Case 4 surprise me, as Case 2 suggested that load_args encoding is not passed to pandas properly, but it is actually used too.

**load_args** seems redundant here.

**Comment by bensdm** [comment]
how to use fs_args for PartitionedDataset?

**Comment by idanov** [comment]
@noklam Does this change c3c378a871305bac30b0575aa8bbc49ee5b72777 that will go in 0.17 solve it for you? If not, could you please provide us with more information about your environment like Python version and conda environment?

**Comment by lorenabalan** [comment]
@noklam I believe this is addressed in 0.17.0, as per commit referenced above. Could you please confirm if that's the case, and if so consider closing the issue?
